googleapiclient/http.py

```diff
@@ -82,22 +86,45 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
     resp, content - Response from the http request (may be HTTP 5xx).
   """
   resp = None
+  content = None
   for retry_num in range(num_retries + 1):
     if retry_num > 0:
-      sleep(rand() * 2**retry_num)
+      sleep_time = rand() * 2 ** retry_num
       logging.warning(
-          'Retry #%d for %s: %s %s%s' % (retry_num, req_type, method, uri,
+          'Sleeping %.2f seconds before retry %d of %d for %s: %s %s%s' %
+          (sleep_time, retry_num, num_retries, req_type, method, uri,
           ', following status: %d' % resp.status if resp else ''))
+      sleep(sleep_time)
 
     try:
       resp, content = http.request(uri, method, *args, **kwargs)
-    except ssl.SSLError:
+    except (ssl.SSLError, TimeoutError) as ex:
       if retry_num == num_retries:
         raise
       else:
+        logging.warning('Encountered exception %s', repr(ex))
         continue
+    reason = ''
+    if resp and resp.status == 403 and content:
+      try:
+        data = json.loads(content.decode('utf-8'))
+        reason = data['error']['errors'][0]['reason']
+      except (ValueError, KeyError):
+        logging.warning('Invalid JSON content from response')
+        try:
+          reason = content.decode('utf-8')
+        except UnicodeDecodeError:
+          reason = str(content)
```

Comment:
I'm unsure on this behavior. If the content can't be decoded as utf-8, it probably doesn't have a useful human-readable string representation. It's probably better to just pass here.